### ISSUE TITLE
Question about Rx Grammar

### ISSUE THREAD

**Issue by 3dGrabber** [dotnet/reactive#1401]
I know that Rx is based on the grammar:

```
OnNext* (OnCompleted | OnError)
```

Is this the only way Rx can work, or is it just a design decision?
I'm asking because I find myself wanting Rx with

```
(OnNext | OnError)* | OnCompleted 
```

My understanding is that this would include the Either monad into Rx (with `left` being the Exception type) and allow us to have [Railway Oriented](https://fsharpforfunandprofit.com/rop/) sequences.

Is this possible in principle?

**Comment by danielcweber** [comment]
Well I would argue that it's the most natural grammar there is. Have a (possibly infinite) amount of values, potentially terminated gracefully or with an error.

Your suggestion seems odd, there's no way for any number of OnNext with graceful termination (OnCompleted). Maybe just a typo.

In any case, you can go kind of meta by materializing your sequence by calling `Materialize`. What you'll get is a sequence of `Notification<T>` where a Notification carries either a value, an exception or nothing (in case of completion). Note that this sequence will of course adhere to standard Rx grammar. Not sure if that is what you need though.

**Comment by bartdesmet** [comment]
There are indeed two possible models here. One is the sequence algebra of `IObservable<T>` which is completely analogous (in fact, the mathematics dual) to `IEnumerable<T>` and that's represented by:

```
OnNext* (OnError | OnCompleted)
```

An alternative is an algebra where there are N channels of different types (e.g. `T`, `Exception`, something else) along with termination. For example, `|` pipes in shell scripting languages such as `PowerShell` where you can have values, error records, progress records, etc. If we just consider errors, that'd be:

```
(OnNext | OnErrror)* OnCompleted
```

which is really isomorphic to

```
(OnNext<T | Exception>)* OnCompleted
```

where the payload of `OnNext` messages carries a variant of either `T` or `Exception` (or whatever else). And then you may as well still throw in a notion of fatal error (e.g. in PowerShell, you can have a `PipelineException` when the pipeline fails for some external reason):

```
(OnNext<T | Exception>)* (OnError | OnCompleted)
```

This was in fact the approach we took when building an Rx to PowerShell bridge (I don't know where that ended up), where the sequence was in fact using a discriminated union of `PSObject` and various other "records" in PowerShell (e.g. errors, progress, and whatever else may flow on the pipeline). The main drawback is that C# doesn't make it easy to consume such variants or discriminated unions, though with expression-based switch, it can become a little easier (and at least fit in an expression):

```csharp
ps.CommandAsObservable("dir").Where(n => n.Kind switch
{
  PsKind.Value => ...,
  PsKind.Error => ...,
  _ => false
}).Subscribe(...);
```

where the "notifications" would have a `Kind` enum property you can switch on. Or, they could all derive from some common notification type and use subtyping to do a type switch on:

```csharp
ps.CommandAsObservable("dir").Where(n => n switch
{
  PsValue v => f(v),
  PsError e => g(e),
  _ => false
}).Subscribe(...);
```

Rx's own discriminated union type to represent `Next | Error | Completed` is, as @danielcweber said, `Notification<T>`, so you could use an `IObservable<Notification<T>>` at the expensive of the code consuming it becoming more tricky.

In very, very early prototype days of Rx, we did have `IEvent<T>` with value and error channels (which could even be swapped for one another in some fancy operators) but it became very tricky rather quickly. Many operators ended up being applicable to either channel (e.g. `Select` on values or on errors, and maybe an error can become a value, i.e. an "error handler" such as `Catch` could be a fancy form of `Select`), which grew the API surface quite badly. It's only when we discovered the duality between `IEnumerable<T>` and `IObservable<T>` that we settled on the natural symmetry between both. And to get the old behavior of possibly wanting to multiplex errors and values over a single "channel", we introduced the reificiation operators, which were originally called `Reify` and `Deify` but got renamed to `Materialize` and `Dematerialize` and that's where `Notification<T>` was born. It's not used that often, but it can be used to enable scenarios like the one you describe, but without language ergonomics and none of the operators is special-cased for `Notification<T>` (other than `Dematerialize<T>`); it's just another object to them.

**Comment by bartdesmet** [comment]
If anyone is interested to re-create a PowerShell-to-Rx bridge, the discussion above could be of help. The entry point in PowerShell APIs to start from is the `Runspace` type (and factories thereof) and the `CreatePipeline` API, if memory serves well.

**Comment by bartdesmet** [comment]
Closing this given it's a question and proper answers have been provided, and there's no issue that needs to be addressed. Feel free to reply to keep the discussion going if anything is unclear.

**Comment by 3dGrabber** [comment]
Sorry for being late...  

Thank you Bart for your explanation and interesting historical perspective.  
I made a little prototype using `Materialize` and `Notification<T>`.  
It works, but it feels clumsy, and interop with ordinary `IObservable`s is annoying.  
Effectively this gives you
```
(Next | Error | Completed) | (OnError | OnCompleted)
```
when you want something much simpler.  
Not to mention the `IObservable<Notification<T>>` noise everywhere.   

> Side note: I would be really helpful if we could do this in C#:   
> `using IMyObservable<T> = IObservable<Notification<T>>`
  
<br>
  
Going through the RX sources here, I realized that internally most of it is centered around [Sink.cs](https://github.com/dotnet/reactive/blob/main/Rx.NET/Source/src/System.Reactive/Internal/Sink.cs).  

Actually, I _think_ one could achieve `(OnNext | OnError)* OnCompleted` behavior by simply removing [this line](https://github.com/dotnet/reactive/blob/d3a4e7269557701164ba56bc56a53ceedd6cb6a0/Rx.NET/Source/src/System.Reactive/Internal/Sink.cs#L61).  

Most operators would probably work just fine with this.  
Only few make strong assumptions about nothing following `OnError`.  
(`OnErrorResumeNext` comes to mind)

I wanted to try this out, but after much trying I gave up, failing to build RX on my computer.
